test/app-luatest/shutdown_test.lua

```diff
@@ -118,3 +118,45 @@ g.test_box_session_new_after_shutdown_started = function(cg)
         server:close()
     end)
 end
+
+local g_snap = t.group('snapshot')
+
+g_snap.before_each(function(cg)
+    cg.server = server:new()
+    cg.server:start()
+end)
+
+g_snap.after_each(function(cg)
+    if cg.server ~= nil then
+        cg.server:drop()
+    end
+end)
+
+-- Test shutdown does not hang due to memtx space snapshot in progress.
+g_snap.test_shutdown_during_memtx_snapshot = function(cg)
+    t.tarantool.skip_if_not_debug()
+    cg.server:exec(function()
+        local fiber = require('fiber')
+        box.schema.create_space('test')
+        box.space.test:create_index('pk')
+        for i=1,10000 do
```

Comment:
Please check that the test doesn't pass without cancellation. I think that the number of records may not be enough to create a large enough delay because xlog records are written in batches. It may be better to inject the delay into `memtx_checkpoint_write_tuple` after all.